### ISSUE TITLE
Problems in the training process

### ISSUE THREAD

**Issue by Pang-b0** [diku-dk/credanno#2]
Hello author, thank you for such a great article. But I found some problems when I reproduced the results. The function (extract_features) seems to have an influence on the result when the code extracts the cluster center. I tried to get the center by clustering in advance and train it directly as indicators, which would be different from calling the function to extract features in the program and then getting the clustering center to train (the clustering center is exactly the same).
![image](https://github.com/diku-dk/credanno/assets/83155460/b6cc9eff-2a86-4197-9579-a99cfaecfd95)

**Comment by ludles** [comment]
The cluster centres are used only to select seed samples. As in 
https://github.com/diku-dk/credanno/blob/ff63a82dbf40908d11e4a7b1f79d2d9ec23305f9/eval_linear_joint_recycle.py#L142-L143

I am sorry that I do not get what you mean by "clustering in advance and train it directly as indicators". Could you elaborate?

**Comment by Pang-b0** [comment]
Thank you for your reply, I mean “If I take this code out alone and run it, the clustering center is indicators = [1414, ..., 500].”，Now that we have the cluster center, in order to simplify our subsequent code running, I will directly assign these recorded values to indices (that is, cancel the code extract_features), which will eventually affect the accuracy.
for example:
We run in the original code.
![image](https://github.com/diku-dk/credanno/assets/83155460/d9353aa2-29b7-447b-a3dd-441be80c20b5)
I got indices in the output.
![image](https://github.com/diku-dk/credanno/assets/83155460/acf5e993-5fc2-46c4-9425-850ae3eba65b)
Since we haven't changed the dino model and data set, the next implementation will definitely be these 19 clustering centers (which I have experimented with).
Therefore, in order to simplify the next execution, we will not perform the function operation of extract_feature. I changed the code to the following example.
![image](https://github.com/diku-dk/credanno/assets/83155460/f724c390-ec30-4836-8be5-8e760ab0bb03)
Then re-running, it is found that the accuracy of the results is affected, both in "seed" and "boost" modes.

**Comment by Pang-b0** [comment]
Thank you for your reply, I mean “If I take this code out alone and run it, the clustering center is indicators = [1414, ..., 500].”，If I take this code out alone and run it, the clustering center is indicators = [1414, ..., 500]. Now that we have the cluster center, in order to simplify our subsequent code running, I will directly assign these recorded values to indices (that is, cancel the code extract_features), which will eventually affect the accuracy.

**Comment by Pang-b0** [comment]
> The cluster centres are used only to select seed samples. As in
> 
> https://github.com/diku-dk/credanno/blob/ff63a82dbf40908d11e4a7b1f79d2d9ec23305f9/eval_linear_joint_recycle.py#L141-L143
> 
thats mean I created a new python file to run the extract_feature part of your code, and I'm sure that the output of indices in the source code is the same as that in the new python file. Then, after annotating the relevant part of "extract_features" in the source code, directly assign the values of indices (the cluster centers extracted in advance in the newly-built python file), which will affect the accuracy, even the difference is 2%.

**Comment by Pang-b0** [comment]
> > The cluster centres are used only to select seed samples. As in
> > https://github.com/diku-dk/credanno/blob/ff63a82dbf40908d11e4a7b1f79d2d9ec23305f9/eval_linear_joint_recycle.py#L140-L141
> 
> thats mean I created a new python file to run the extract_feature part of your code, and I'm sure that the output of indices in the source code is the same as that in the new python file. Then, after annotating the relevant part of "extract_features" in the source code, directly assign the values of indices (the cluster centers extracted in advance in the newly-built python file), which will affect the accuracy, even the difference is 2%.

I think the extract_features function in line 141 of the code may have affected the parameters in the model, but I saw that you set "model.eval ()", so I'm not sure why it has the influence of precision.

**Comment by ludles** [comment]
I have to admit that I have not tried this approach myself. To locate the problem, I would suggest trying [`torch.allclose()`](https://pytorch.org/docs/stable/generated/torch.allclose.html#torch-allclose) on `train_features`, `km.cluster_centers_`, etc., to narrow down the places of potential introduced randomness. 

Randomness is a known issue in this problem. Different random seeds can certainly lead to fluctuation in performance. It actually makes more sense to run several it several times and compare the statistical mean and variance in performance.

**Comment by Pang-b0** [comment]
> I've tested it many times and it's still these seeds. That is, indices has not changed.

**Comment by Pang-b0** [comment]
> 我不得不承认，我自己还没有尝试过这种方法。为了定位问题，我建议尝试 [](https://pytorch.org/docs/stable/generated/torch.allclose.html#torch-allclose) 、 等，以缩小潜在引入随机性的位置。`train_features``km.cluster_centers_`
> 
> 随机性是此问题中的一个已知问题。不同的随机种子肯定会导致性能波动。实际上，多次运行几次并比较统计平均值和性能方差更有意义。

Do you mean that train_features will affect the subsequent training? But what I understand is that if indices remains unchanged, there should be no such influence.

**Comment by Pang-b0** [comment]
I don't know if it's my improper operation. I found that after my dino model is fixed, if the parameters are consistent, the result will be the same.
For example, I executed the code below twice without modifying the source code, and their output precision indicators were exactly the same.
![image](https://github.com/diku-dk/credanno/assets/83155460/ff9f1e01-02b2-46aa-ba57-f2bd0bc34d65)
